nncf/torch/extensions/src/quantization/cuda/functions_cuda_impl.cu

```diff
@@ -352,20 +360,31 @@ std::vector<at::Tensor> q_single_scale_cuda_backward(at::Tensor grad_output,
 
     auto grad_input_range = at::empty({1}, grad_output.options());
     auto grad_input_low = at::empty({1}, grad_output.options());
+    
 
     auto grid_size = std::min(GET_BLOCKS(size), CUDA_BLOCKS_PER_GRID_FOR_UNIFORM_ELTWISE);
     auto dev_tmp_range = at::empty({grid_size}, grad_output.options());
     auto dev_tmp_low = at::empty({grid_size}, grad_output.options());
     auto dev_last_block_counter_range = at::zeros({1},  at::device(grad_output.options().device()).dtype(at::kInt));
     auto dev_last_block_counter_low = at::zeros({1},  at::device(grad_output.options().device()).dtype(at::kInt));
 
+    const auto grad_input_dtype = grad_output.scalar_type();
+    if (grad_input_dtype == c10::ScalarType::Half)
+    {
+        grad_input_range = grad_input_range.toType(c10::ScalarType::Float);
+        grad_input_low = grad_input_low.toType(c10::ScalarType::Float);
+        dev_tmp_range = dev_tmp_range.toType(c10::ScalarType::Float);
+        dev_tmp_low = dev_tmp_low.toType(c10::ScalarType::Float);
+    }
```

Comment:
`toType` method is creating new tensor and is copying casted values from a source tensor.
 
Maybe if we will keep FP16 memory of `grad_input_range` and `grad_output_range` before casting and then copy result from copy FP32 values of `grad_input_range` and `grad_output_range` back to FP16 we will safe 2 calls of Tensor constructor/destructor.

Reply:
CC: @vshampor, @alexsu52